packages/spawner/test/swingsetTests/contractHost/test-contractHost.js

```diff
@@ -89,7 +73,8 @@ const contractBobFirstGolden = [
   'bob stock balance {"label":{"issuer":{},"allegedName":"fudco"},"extent":1996}',
 ];
 
-test.skip('run contractHost Demo --bob-first with SES', async t => {
+// stuck on https://github.com/Agoric/agoric-sdk/issues/615#issuecomment-592060759 for now
+test('run contractHost Demo --bob-first with SES', async t => {
```

Comment:
The current state of the PR is that this test is the only one to run and fails 
Right now, it fails with 
```sh
FAILED ASSERTION false
payment not found: Promise <Promise <[Object: null prototype] {}>> [Promise] { <pending> }
    doProcess: vat[v4][o+4].getAmountOf dispatch failed: Error <Object <[Object: null prototype] {}>>: payment not found: (an object)
```

in issuer.js, if i rewrite `getAmountOf: paymentLedger.get` to 
```js
getAmountOf(paymentP){
    const payment = E.unwrap(paymentP);
    return paymentLedger.get(payment)
}
```
, i get :
```sh
doProcess: vat[v4][o+4].getAmountOf dispatch failed: TypeError <Object <Object <[Object: null prototype] {}>>>: Value is a Thenble but not a HandledPromise fulfilled to a presence
```

i don't know enough about how E/makePromise/SwingSet work to meaningfully dig into what's going on

Reply:
[ping](https://github.com/Agoric/agoric-sdk/issues/615#issuecomment-599738948) @michaelfig

Reply:
The unwrap can fail if the promise came from a different vat.  In that case, you should probably catch the exception and try awaiting the promise that failed.

Reply:
My understanding of your response is to do something like:

```js
getAmountOf(paymentP){
    let payment;
    try{
        payment = E.unwrap(paymentP);
    }
    catch(e){
        payment = await paymentP
        // or maybe payment = E.unwrap(await paymentP) ?
    }

    return paymentLedger.get(payment)
}
```

and whichever solution it is, it will probably need to be applied to every single argument sent to every single exposed method of ERTP and this doesn't feel like the right place for this code
It's not clear to me whether or how much ERTP code should be aware of and adapt to whether its arguments are the correct objects, promises or promises from another vat

Adding @katelynsills to the discussion for her opinion

Reply:
I may have been wrong.  Is it the case that a valid payment should always be created within the vat that contains the `getAmountOf`?

If so, you can just return `undefined` in the catch block (with no await at all).

Reply:
>  Is it the case that a valid payment should always be created within the vat that contains the getAmountOf?

The answer is "yes" in the sense that currently, `paymentLedger.get` will only accept a valid payment object (and fails an assertion with `payment not found` otherwise). And i think it's a good decision.

What's not clear to me is who's responibility it should be to make sure `paymentLedger.get` gets only valid payment objects
It can be ERTP's (which is what i wrote above and am not convinced with), it can be the "caller". I see two cases for the caller:
- either the caller is code that runs in the same vat and `getAmountOf(payment)` should work as expected. If, for some reason `payment` is a promise in the same vat, the caller can take the responsibility to `await` it and everything should be fine
- or the issuer and the caller are in different vats in which case, the caller does something like `E(issuerP).getAmountOf(paymentP)` (where, on the caller side, `issuerP` and `paymentP` are "presences" representing objects which "real belonging" is in the issuer vat)
and in this case, a couple things are not clear to me (because i lack the knowledge):
    - is a presence a promise?
    - can a presence represent a promise in another vat? (or should it always represent a non-promise object)
    - I would anticipate that `E(issuerP).getAmountOf(paymentP)` (and eventually `issuerP~.getAmountOf(paymentP)`) in one vat would resolve automatically to `issuer.getAmountOf(payment)` in the issuer vat (and consequently no need to play around with `await` or `E.unwrap`), but maybe that there is a reason why it's not possible at all because of language reasons or currently because of temporary limitations

Reply:
Let's move this conversation to https://github.com/Agoric/agoric-sdk/issues/708 to keep the topics together. I will respond there. Thank you for finding this, David!